dashboard/modules/serve/serve_head.py

```diff
@@ -14,81 +13,130 @@
+                    text=resp_text,
+                    status=resp.status,
+                    content_type=resp.content_type,
+                )
+        except Exception:
+            import traceback
+
             return Response(
                 status=503,
                 text=(
-                    "Failed to hit serve agent. "
-                    "Check the dashboard_agent logs to see "
-                    f"if the agent failed to launch. {e}"
+                    f"Failed to hit serve agent at address {url} on "
+                    "the head node. Check the dashboard_agent logs to see "
+                    "if the agent failed to launch. "
+                    f"See traceback:\n{traceback.format_exc()}"
                 ),
             )
 
-    async def _get_head_agent(self) -> Optional[str]:
-        """
-        Grabs the head node dashboard_agent's address.
-        """
-        # the number of agents which has an available HTTP port.
-        raw_agent_infos = await DataOrganizer.get_all_agent_infos()
-        agent_infos = {
-            key: value
-            for key, value in raw_agent_infos.items()
-            if value.get("httpPort", -1) > 0
-        }
-        if not agent_infos:
-            return None
-
-        # TODO(aguo): Get the head agent by node_id instead of node_ip once
-        # head_node_id is easily available in dashboard_head.py
-        head_node_ip = self._dashboard_head.ip
-        logger.info(f"head_node_ip: {head_node_ip}")
-        for agent_info in agent_infos.values():
-            logger.info(f"agent_info: {agent_info}")
-            if agent_info["ipAddress"] == head_node_ip:
-                node_ip = agent_info["ipAddress"]
-                http_port = agent_info["httpPort"]
-                return f"http://{node_ip}:{http_port}"
-
-        return None
+    @routes.get("/api/serve_head/version")
+    async def get_version(self, req: Request) -> Response:
+        return await self.proxy_request(
+            req=req,
+            route="/api/ray/version",
+            method=RestMethod.GET,
+        )
+
+    @routes.get("/api/serve_head/deployments/")
+    @dashboard_optional_utils.init_ray_and_catch_exceptions()
+    async def get_all_deployments(self, req: Request) -> Response:
+        return await self.proxy_request(
+            req=req,
+            route="/api/serve/deployments/",
+            method=RestMethod.GET,
+        )
+
+    @routes.get("/api/serve_head/applications/")
+    @dashboard_optional_utils.init_ray_and_catch_exceptions()
+    async def get_serve_instance_details(self, req: Request) -> Response:
+        return await self.proxy_request(
+            req=req,
+            route="/api/serve/applications/",
+            method=RestMethod.GET,
+        )
+
+    @routes.get("/api/serve_head/deployments/status")
+    @dashboard_optional_utils.init_ray_and_catch_exceptions()
+    async def get_all_deployment_statuses(self, req: Request) -> Response:
```

Comment:
ditto

Reply:
I don't want to make any behavior changes in this PR. I'll leave it as-is since the goal is to mirror the Serve agent.